File: Leet 3090/solution.cs
class Solution
{
    public static int MaximumLengthSubstring(string s)
    {
        var frequency = new Dictionary<char, int>();
        int maxLength = 0;
        int left = 0;

        for (int right = 0; right < s.Length; right++)
        {
            if (!frequency.TryGetValue(s[right], out int value))
            {
                frequency[s[right]] = 0;
            } 
        
            frequency[s[right]] = value + 1;
            
            while (frequency[s[right]] > 2)
            {
                frequency[s[left]]--;
                left++;
            }

            maxLength = Math.Max(maxLength, right - left + 1);
        }

        return maxLength;
    }
}

class Program
{
    private static void TestCase(string s, int expected)
    {
        string result = Solution.MaximumLengthSubstring(s) == expected ? "Pass" : "Fail";
        Console.WriteLine($"Test case should equal {expected} ? {result}");
    }

    private static void Main()
    {
        Console.WriteLine("3090. Maximum Length Substring With Two Occurrences");

        TestCase("bcbbbcba", 4);
        TestCase("aaaa", 2);
    }
}

File: Leet 3090/solution.ts
/**
 * Return the maximum length of a substring that contains at most 2 occurrences of each character.
 * @param s String to check
 * @returns Length of maximum substring.
 */
function maximumLengthSubstring(s: string): number {
  const fr = new Map<string, number>();

  let maxLen: number = 0;
  let left: number = 0;

  for (let right: number = 0; right < s.length; right++) {

    fr.set(s[right], (fr.get(s[right]) ?? 0) + 1);

    while (fr.get(s[right])! > 2) {
      fr.set(s[left], fr.get(s[left])! - 1);
      left++;
    }

    maxLen = Math.max(maxLen, right - left + 1);
  }

  return maxLen;
}

function testCase(s: string, expected: number): void {
  const result: string =
    maximumLengthSubstring(s) === expected ? "Pass" : "Fail";
  console.log(`Test case should equal expected ${expected} ? ${result}`);
}

function main(): void {
  console.log("3090. Maximum Length Substring With Two Occurrences");

  testCase("bcbbbcba", 4);
  testCase("aaaa", 2);
}

main();
